### PR TITLE
Add Declarative UI builders for Python

### PR DESCRIPTION
**Pull request by twojtasz**

### CODE REVIEW

File: python/xviz_avs/builder/xviz_ui_builder.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            "select": XVIZSelectBuilder,
+            "table": XVIZTableBuilder,
+            "treetable": XVIZTreeTableBuilder,
+            "video": XVIZVideoBuilder
+        }
+
+    def child(self, panel):
+        if not isinstance(panel, XVIZPanelBuilder):
+            raise TypeError("Argument panel must be an XVIZPanelBuilder instance")
+        self._panels.append(panel)
```
Comment (jlisee) [inline python/xviz_avs/builder/xviz_ui_builder.py:33]: another `self` return?

File: python/xviz_avs/builder/declarative_ui/panel.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        if self._name is None:
+            self._validateError('Panel must have a "name" set')
+
+    @property
+    def name(self):
+        return self._name
+
+    def child(self, child):
+        self._children.append(child)
+        return child
```
Comment (jlisee) [inline python/xviz_avs/builder/declarative_ui/panel.py:25]: A `self` return?

File: python/xviz_avs/builder/declarative_ui/constants.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class UI_TYPES:
```
[the remark this answers is not in the document]
Comment (twojtasz) [reply python/xviz_avs/builder/declarative_ui/constants.py:1]: I think so, I will look deeper into the Enum support

File: python/xviz_avs/builder/declarative_ui/container.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+        self._validate()
+
+    def _validate(self):
+        if self._name is None:
+            self._validateError('Panel must have a "name" set')
+
+    def child(self, child):
+        self._children.append(child)
+        return child
```
[the remark this answers is not in the document]
Comment (twojtasz) [reply python/xviz_avs/builder/declarative_ui/container.py:20]: This is the same for all of these, this is modeling from the JS code, but I honestly don't understand it.   I am going to review the JS code usages and see if I can understand why we do this.

**Comment by twojtasz** [comment]
@jlisee Thanks for reviewing.  Do you recall why we don't have protobuf defs for the decl UI and just use Struct?

The parity with the JS code for the decl ui builders may break, or change a little.

Mainly, it feels like it doesn't quite match the fluent API of the primitives very well, and I'm not even sure it makes sense to since primitives are individual units generally, but the UI a nesting of components.  I feel the API is awkward as it stands and want to review options.

Obviously i'll probably just leave this work in place and put thoughts together for an alternative once I get a feel.

**Comment by jlisee** [comment]
> @jlisee Thanks for reviewing. Do you recall why we don't have protobuf defs for the decl UI and just use Struct?

It was because changing the declarative UI to be protobuf compatible would of been a lot of work, and I did not consider it worth it at the time.  I was also worried the UX would change too much given protobufs limitations.

> The parity with the JS code for the decl ui builders may break, or change a little.
> 
> Mainly, it feels like it doesn't quite match the fluent API of the primitives very well, and I'm not even sure it makes sense to since primitives are individual units generally, but the UI is more of a nesting of components. I feel the API is awkward as it stands and want to review options.
> 
> I'll probably just leave this work in place and put thoughts together for an alternative once I get a feel.

Yes I think starting with what JS has if that is how it works is a good way to go to make the docs similar.

**Comment by twojtasz** [comment]
@jlisee When testing enums (with string values) they break when converting to the protobuf Struct Type, it doesn't seem to handle Enums currently.

I'm talking to uber to get access to the buildkite details.

**Pull request merged**